cirq/ops/controlled_gate.py

```diff
@@ -12,52 +12,109 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Union
+from typing import Any, List, Union
 
 import numpy as np
 
 import cirq
 from cirq import linalg, protocols, value
-from cirq.ops import raw_types, controlled_operation as cop
+from cirq.ops import (raw_types, controlled_operation as cop, gate_operation)
 from cirq.type_workarounds import NotImplementedType
 
 
 @value.value_equality
 class ControlledGate(raw_types.Gate):
     """Augments existing gates with a control qubit."""
 
-    def __init__(self, sub_gate: raw_types.Gate) -> None:
+    def __init__(self, sub_gate: raw_types.Gate,
+                 control_qubits: List[raw_types.Qid] = None,
+                 num_controls: int = 1) -> None:
         """Initializes the controlled gate.
 
         Args:
             sub_gate: The gate to add a control qubit to.
+            control_qubits: The qubits that would act as controls.
+            num_control: Total number of control qubits.
         """
-        self.sub_gate = sub_gate
+        if control_qubits is None:
+            control_qubits = []
+        if num_controls < len(control_qubits):
+            raise ValueError('More specified control qubits than num_controls')
+
+        # Leave unspecified controls as Nones.
+        self.control_qubits = ([None]*(num_controls - # type: ignore
+                                       len(control_qubits)) +
+                               control_qubits)
+
+        self.num_controls = num_controls
```

Comment:
This is redundant w.r.t. the length of the list. Don't store it.

Reply:
Removed. Converted to a method.